test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java

```diff
@@ -329,70 +343,94 @@ public void testPyScikitKMeans() throws Exception {
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(getInfoEKN.get(0).getE7()  + "/" + getInfoEKN.get(0).getE1() + "/" +  getInfoEKN.get(0).getE5(), getId);
+
         FeatureClusters clSet = res.getData().asClassInstance(FeatureClusters.class);
         checkKMeansForSample(clSet);
     }
-    
-    @Test
+
     public void testCorrectMatrix() throws Exception {
         String testWsName = getWsName();
         String sourceMatrixId = "notcorrected_matrix.1";
+
+        FloatMatrix2D testmat = getSampleMatrix();
         ExpressionMatrix data = new ExpressionMatrix().withType("log-ratio").withScale("1.0")
-                .withData(getSampleMatrix());
+                .withData(testmat);
+        //introduce null value for imputation test
         data.getData().getValues().get(0).set(0, null);
         Assert.assertEquals(1, getNullCount(data.getData()));
-        getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
+        List<Tuple11<Long, String, String, String, Long, String, Long, String, String, Long, Map<String, String>>> infoCM = getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
                 new ObjectSaveData().withName(sourceMatrixId).withType("KBaseFeatureValues.ExpressionMatrix")
-                .withData(new UObject(data)))));
+                        .withData(new UObject(data)))));
         String targetMatrixId = "corrected_matrix.1";
-        impl.correctMatrix(new CorrectMatrixParams().withInputData(
+        String getId = impl.correctMatrix(new CorrectMatrixParams().withInputData(
                 testWsName + "/" + sourceMatrixId).withOutWorkspace(testWsName)
                 .withOutMatrixId(targetMatrixId).withTransformType("missing"), token, getContext());
+
         ExpressionMatrix matrix = getWsClient().getObjects(Arrays.asList(
                 new ObjectIdentity().withWorkspace(testWsName).withName(targetMatrixId)))
                 .get(0).getData().asClassInstance(ExpressionMatrix.class);
+
+        //test matrix imputation
         Assert.assertEquals(0, getNullCount(matrix.getData()));
-        Assert.assertEquals(0.325, (double)matrix.getData().getValues().get(0).get(0), 1e-10);
+        Assert.assertEquals(0.325, (double) matrix.getData().getValues().get(0).get(0), 1e-10);
+
+        //reload with returned object id
+        ExpressionMatrix res = loadfromWs(getId).asClassInstance(ExpressionMatrix.class);
+        //reset from null in data before imputation
+        data.getData().getValues().get(0).set(0, res.getData().getValues().get(0).get(0));
+        Assert.assertEquals(data.getData().toString(), matrix.getData().toString());
+
+        //test if returned id matches id returned by ws saveobjects
+        Assert.assertEquals(infoCM.get(0).getE7()  + "/" + infoCM.get(0).getE1() + "/" +  infoCM.get(0).getE5(), getId);
```

Comment:
Why is the method returning the ref for an object that was created prior to the method being run? Doesn't the method do anything?

Reply:
There's more cases of this, I won't mark them

Reply:
this is now fixed -- the object id has to be incremented in this case

Reply:
and the few other cases

Reply:
Now I'm confused as to how the tests passed previously - the method behavior didn't change at all AFAICT but the tests passed, but now the tests add 1 to the object ID or version and the tests still pass. How is that possible?

Reply:
ha, it turns out that these tests did not pass as I think the classes got cached at some point along the way ... but pass now with the +1 increments.